CodeListLibrary_project/clinicalcode/permissions.py

```diff
@@ -22,7 +35,67 @@ class Permissions:
 
     PERMISSION_CHOICES_WORLD_ACCESS = ((NONE, 'No Access'), (VIEW, 'View'))
 
+def get_entity_history_id(entity):
+    """
+        Tries to get the entity's history id through reflection.
 
+        @param entity, an entity of the model
+        
+        @returns int, a history id
+    """
+
+    set_cls = entity._meta.model_name
+    entity = entity.history.latest()
+    if set_cls == 'publishedconcept':
+        return int(entity.concept_history_id)
+    elif set_cls == 'publishedphenotype':
+        return int(entity.phenotype_history_id)
+    elif set_cls == 'publishedworkingset':
+        return int(entity.workingset_history_id)
+    return None
+
+def try_get_valid_history_id(set_cls, request, pk):
+    """
+        Tries to resolve a valid history id for an entity query.
+
+        If the user is a superuser, the owner, or is a moderator this method will return the most recent version regardless of publication status.
+
+        Otherwise, this method will return the most recently published version, if available.
+
+        @param set_cls, a model
+        @param request, the request
+        @param pk, the id of the entity
+
+        @returns int, a history_id
+
+    """
+    
+    from .models.Concept import Concept
+    from .models.Phenotype import Phenotype
+    from .models.PhenotypeWorkingset import PhenotypeWorkingset
+    from .models.PublishedConcept import PublishedConcept
+    from .models.PublishedPhenotype import PublishedPhenotype
+    from .models.PublishedWorkingset import PublishedWorkingset
+
+    user = request.user
+    entity = set_cls.objects.get(pk=pk)
+    
+    if user.is_superuser or entity.owner == user or is_member(user, "Moderators"):
```

Comment:
this needs to be through permission as we have other conditions for access

Reply:
This is there just to determine whether to traverse through the historical records, the actual permision checks occur within the phenotype/concept/phenotypeworkingset views through the call to the method ```validate_access_to_view``` on lines 373, 188 and 579 respectively.
 
I guess there is a possibility that some cases may not be caught that may require traversal, so I added ```allowed_to_view``` method calls into this PR though as you requested